File: frontend/src/types/kubernetes.ts
/**
 * The deliberately small, read-only contract served by `/kubernetes/topology`.
 *
 * These are observed Kubernetes resources, not editable Homelable canvas nodes.
 * Keep this contract free of labels, annotations, manifests, and credentials so
 * the same payload is safe to use in the visual view and the agent-facing MCP
 * resource.
 */
export type KubernetesSyncState =
  | 'disabled'
  | 'never_synced'
  | 'syncing'
  | 'fresh'
  | 'stale'
  | 'error'

export interface KubernetesCluster {
  id: string
  name: string
}

export interface KubernetesSync {
  state: KubernetesSyncState
  last_success_at?: string | null
  last_attempt_at?: string | null
  last_error?: string | null
  object_count?: number
  relationship_count?: number
}

export interface KubernetesTopologyObject {
  id: string
  kind: string
  name: string
  namespace?: string | null
  status?: string | null
  /** A bounded set of safe summary facts (for example ingress host/path/port). */
  properties?: Record<string, unknown> | null
}

export interface KubernetesTopologyRelationship {
  source: string
  target: string
  kind: 'contains' | 'owns' | 'scheduled_on' | 'routes_to' | 'has_endpoint' | string
  properties?: Record<string, unknown> | null
}

export interface KubernetesTopology {
  schemaVersion: number
  cluster: KubernetesCluster
  sync: KubernetesSync
  objects: KubernetesTopologyObject[]
  relationships: KubernetesTopologyRelationship[]
}

export interface KubernetesStatus extends KubernetesSync {
  enabled: boolean
  configured: boolean
}

File: backend/app/schemas/kubernetes.py
"""Public, sanitized Kubernetes topology response schemas."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

KubernetesTopologyState = Literal["disabled", "never_synced", "syncing", "fresh", "stale", "error"]


class KubernetesSyncOut(BaseModel):
    enabled: bool
    configured: bool
    state: KubernetesTopologyState
    last_success_at: datetime | None = None
    last_attempt_at: datetime | None = None
    last_error: str | None = None
    object_count: int = 0
    relationship_count: int = 0


class KubernetesClusterOut(BaseModel):
    id: str
    name: str


class KubernetesObjectOut(BaseModel):
    id: str
    kind: str
    name: str
    namespace: str | None = None
    status: str | None = None
    properties: dict[str, Any] | None = None


class KubernetesRelationshipOut(BaseModel):
    source: str
    target: str
    kind: str
    properties: dict[str, Any] | None = None


class KubernetesTopologyOut(BaseModel):
    schema_version: int = Field(1, serialization_alias="schemaVersion")
    cluster: KubernetesClusterOut
    sync: KubernetesSyncOut
    objects: list[KubernetesObjectOut]
    relationships: list[KubernetesRelationshipOut]
